evm/p2p/lightchain.py

```diff
@@ -1,69 +1,270 @@
+                kademlia.Node(
+                    keys.PublicKey(decode_hex("584c0db89b00719e9e7b1b5c32a4a8942f379f4d5d66bb69f9c7fa97fa42f64974e7b057b35eb5a63fd7973af063f9a1d32d8c60dbb4854c64cb8ab385470258")),  # noqa: E501
+                    kademlia.Address("51.15.35.2", 30303, 30303)),
+                kademlia.Node(
+                    keys.PublicKey(decode_hex("d40871fc3e11b2649700978e06acd68a24af54e603d4333faecb70926ca7df93baa0b7bf4e927fcad9a7c1c07f9b325b22f6d1730e728314d0e4e6523e5cebc2")),  # noqa: E501
+                    kademlia.Address("51.15.132.235", 30303, 30303)),
+                kademlia.Node(
+                    keys.PublicKey(decode_hex("482484b9198530ee2e00db89791823244ca41dcd372242e2e1297dd06f6d8dd357603960c5ad9cc8dc15fcdf0e4edd06b7ad7db590e67a0b54f798c26581ebd7")),  # noqa: E501
+                    kademlia.Address("51.15.75.138", 30303, 30303)),
+            ]
+        else:
+            raise ValueError("Unknown network_id: %s", self.network_id)
 
 
 class LightChain(Chain):
-    on_demand_data_backend_class = None
+    # FIXME: Should be passed in by callers
+    privkey = ecies.generate_privkey()
+    # TODO:
+    # 1. Implement a queue of requests and a distributor which picks items from that queue and
+    # sends them to one of our peers, ensuring the selected peer has the info we want,
+    # retrying on timeouts and respecting the flow control rules.
 
-    def __init__(self, chaindb, header=None):
-        super(LightChain, self).__init__(chaindb, header=header)
-        self.on_demand_data_backend = self.on_demand_data_backend_class(self.chaindb)
+    def __init__(self, chaindb):
+        super(LightChain, self).__init__(chaindb)
+        self.peer_pool = PeerPool(chaindb, self.network_id, self.privkey)
+
+    async def run(self):
+        await self.peer_pool.run()
 
     async def stop(self):
-        await self.on_demand_data_backend.stop()
+        self.logger.info("Stopping ...")
+        await self.peer_pool.stop()
 
     async def get_canonical_block_by_number(self, block_number):
         try:
             block_hash = self.chaindb.lookup_block_hash(block_number)
         except KeyError:
-            raise BlockNotFound("No block with number {} found on canonical chain".format(
-                block_number))
+            raise BlockNotFound(
+                "No block with number %s found on local chain", block_number)
         return await self.get_block_by_hash(block_hash)
 
+    @alru_cache(maxsize=1024)
     async def get_block_by_hash(self, block_hash):
+        peer = await self.peer_pool.get_best_peer()
+        self.logger.debug("Fetching block %s from %s", encode_hex(block_hash), peer)
+        request_id = gen_request_id()
+        peer.les_proto.send_get_block_bodies([block_hash], request_id)
+        reply = await peer.wait_for_reply(request_id)
+        if len(reply['bodies']) == 0:
+            raise BlockNotFound("No block with hash %s found", block_hash)
```

Comment:
Same here with formatting of the exception message.

Reply:
Reverted as well